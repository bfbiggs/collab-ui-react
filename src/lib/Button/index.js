import React from 'react';
import PropTypes from 'prop-types';
import { Loading } from '@collab-ui/react';

/**
 * @category controls
 * @component button
 * @variations collab-ui-react
 */

class Button extends React.PureComponent {
  static displayName = 'Button';

  componentDidMount() {
    const { ariaLabel, ariaLabelledBy } = this.props;
    /* eslint-disable no-console */
    (!ariaLabel && !ariaLabelledBy)
      &&
      console.warn('Accessibility could be improved with ariaLabel');
  }

  handleKeyPress = e => {
    const { onClick } = this.props;
    e.preventDefault();
    if (e.which === 32 || e.which === 13) {
      return onClick(e);
    } else if (e.charCode === 32 || e.charCode === 13) {
      return onClick(e);
    }
  };

  render() {
    const {
      active,
      label,
      color,
      expand,
      onClick,
      ariaLabel,
      ariaLabelledBy,
      className,
      disabled,
      loading,
      large,
      style,
      type,
      tag,
      href,
      circle,
      children,
      ...otherHTMLProps
    } = this.props;

    const getChildren = () => {
      return (
        [
          <div style={{ opacity: `${loading ? 1 : 0}` }} key='child-0'>
            <Loading
              small={circle && !large} />
          </div>,
          <span
            className='cui-button__children'
            style={{ opacity: `${loading ? 0 : 1}` }}
            key='child-1'
          >
            {children}
          </span>
        ]
      );
    };

    const button = React.createElement(
      tag,
      {
        ref: 'button',
        className:
          'cui-button' +
          `${(circle && ` cui-button--circle`) || ''}` +
          `${(large && ` cui-button--large`) || ''}` +
          `${(expand && ` cui-button--expand`) || ''}` +
          `${(color && ` cui-button--${color}`) || ''}` +
          `${(active && ` active`) || ''}` +
          `${(className && ` ${className}`) || ''}`,
        onClick: onClick,
        onKeyPress: this.handleKeyPress,
        style: style,
        disabled: disabled || loading,
        alt: ariaLabel || label,
        href: (tag === 'a' && href) || undefined,
        type: tag !== 'a' && type || '',
        role: (tag !== 'button' && 'button') || '',
        ...ariaLabel
          ? { 'aria-label': ariaLabel }
          : { 'aria-labelledby': ariaLabelledBy },
        tabIndex: 0,
        ...otherHTMLProps,
      },
      getChildren()
    );

    return (
      label
        ?
        <div
          className={`cui-button__container${large ? '' : '--small'}`}
        >
          {button}
          <div className='cui-button__label'>
            {label}
          </div>
        </div>
        :
        button
    );
  }
}

Button.propTypes = {
  /**
   * Text to display inside the button
   */
  label: PropTypes.string,
  /**
   * Children Nodes to Render inside button
   */
  children: PropTypes.node.isRequired,
  /**
   * optional circle prop type
   */
  circle: PropTypes.bool,
  /**
 * optional active prop type
 */
  active: PropTypes.bool,
  /**
   * optional large prop type
   */
  large: PropTypes.bool,
  /**
   * optional color prop type
   */
  color: PropTypes.string,
  /**
   * optional tag prop type
   */
  tag: PropTypes.oneOf(['button', 'input', 'a']),
  /**
   * optional type prop type
   */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  /**
   * Handler to be called when the user taps the button
   */
  onClick: PropTypes.func,
  /**
   * Href prop changes element to anchor element
   */
  href: PropTypes.string,
  /**
   * Text to display for blindness accessibility features
   */
  ariaLabel: PropTypes.string,
  /**
 * ID to reference for blindness accessibility features
 */
  ariaLabelledBy: PropTypes.string,
  /**
   * @ignore
   * optional css class string
   */
  className: PropTypes.string,
  /**
   * Sets the attribute expanded to the button
   */
  expand: PropTypes.bool,
  /**
   * Sets the attribute disabled to the button
   */
  disabled: PropTypes.bool,
  /**
   * Activates the loading animation and sets the button to disabled
   */
  loading: PropTypes.bool,
  /**
   * Additional styling applied to the button
   * @ignore
   */
  style: PropTypes.object,
};

Button.defaultProps = {
  active: false,
  ariaLabel: '',
  ariaLabelledBy: '',
  expand: false,
  large: false,
  className: '',
  type: 'button',
  color: '',
  disabled: false,
  loading: false,
  style: {},
  onClick: null,
  tag: 'button',
  label: '',
  children: null,
  href: '',
  circle: false,
};

export default Button;

/**
* @name Default Buttons
* @description Default Button.
*
* @category controls
* @component button
* @section default
*
* @js

export default function ButtonDefault() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div className='columns small-3'>
        <Button
          children='Test Me'
          onClick={() => {}}
          ariaLabel='For the Win'
        />
      </div>
    </div>
  );
}

**/

/**
* @name Large Buttons
* @description Create large buttons by passing in the <code>large</code> prop.
*
* @category controls
* @component button
* @section large
*
* @js
*
export default function ButtonDefault() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div className='columns small-3'>
        <Button
          children='Test Me'
          onClick={() => {}}
          ariaLabel='For the Win'
          large
        />
      </div>
    </div>
  );
}
*/

/**
* @name Button Tags
* @description Button with Tag
*
* @category controls
* @component button
* @section tags
*
*
* @js
export default function ButtonTags() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div>Tag Attribute (a)</div>
      <div className='columns small-3'>
        <Button
          children='Link'
          onClick={() => {}}
          ariaLabel='For the Win'
          tag='a'
        />
      </div>
    </div>
  );
}
**/

/**
* @name Button with Color
* @description Button with Color
*
* @category controls
* @component button
* @section colors
*
*
* @js
export default function ButtonColors() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div>Color Attribute</div>
      <div className='columns small-3'>
        <Button
          children='Test Me'
          onClick={() => {}}
          ariaLabel='For the Win'
          color='blue'
        />
      </div>
    </div>
  );
}
**/

/**
* @name Expanded Buttons
* @description Create block level buttons —those that span the full width of a parent— by passing in the <code>expand</code> prop.
*
* @category controls
* @component button
* @section expanded
*
*
* @js

export default function ButtonExpanded() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div className='columns small-3'>
        <Button
          children='Test Me'
          onClick={() => {}}
          ariaLabel='For the Win'
          expand
        />
      </div>
    </div>
  );
}
**/

/**
* @name Disabled Buttons
* @description Create large buttons by passing in the <code>disabled</code> prop.
*
* @category controls
* @component button
* @section disabled
*
*
* @js

export default function ButtonDisabled() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
    <div className='columns small-3'>
      <Button
        children='Test Me'
        onClick={() => {}}
        ariaLabel='For the Win'
        disabled
      />
    </div>
    </div>
  );
}
**/

/**
* @name Loading Buttons
* @description Create loading buttons by passing in the <code>loading</code> prop.
*
* @category controls
* @component button
* @section loading
*
*
* @js

export default function ButtonLoading() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div className='columns small-3'>
        <Button
          children='Test Me'
          onClick={() => {}}
          ariaLabel='For the Win'
          loading
        />
      </div>
    </div>
  );
}
**/

/**
* @name Button with circle
* @description Button with circle, default is small
*
* @category controls
* @component button
* @section circle
*
*
* @js
import { Icon } from '@collab-ui/react';

export default function ButtonShape() {
  return(
    <div className='row' style={{marginBottom: '1rem'}}>
      <div className='columns small-3'>
        <Button
          children={<Icon name='icon-search_12' />}
          onClick={() => {}}
          ariaLabel='For the Win'
          circle
        />
      </div>
    </div>
  );
}
**/

/**
* @name Circular Buttons
* @description Create circular buttons by passing in the <code>circle</code> prop.
*
* @category controls
* @component button
* @section circle
*
*
* @js
import { Icon } from '@collab-ui/react';

export default function ButtonLargeCircle() {
  return(
    <div>
      <div className='row' style={{marginBottom: '1rem'}}>
        <div>Circle Button</div>
        <div className='columns small-3'>
          <Button
            children={<Icon name='icon-search_12' />}
            onClick={() => {}}
            ariaLabel='For the Win'
            circle
          />
        </div>
      </div>
      <div className='row' style={{marginBottom: '1rem'}}>
        <div>Large Circle Button</div>
        <div className='columns small-3'>
          <Button
            children={<Icon name='icon-search_16' />}
            onClick={() => {}}
            ariaLabel='For the Win'
            circle
            large
          />
        </div>
      </div>
    </div>
  );
}
**/
