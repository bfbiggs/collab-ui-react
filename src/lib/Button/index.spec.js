import React from 'react';
import { shallow, mount } from 'enzyme';
import Button from '../Button';
import { Loading } from '@collab-ui/react';

describe('tests for <Button />', () => {
  it('should match SnapShot', () => {
    const container = shallow(<Button children='test' ariaLabel='test' />);

    expect(container).toMatchSnapshot();
  });

  it('should render one Button', () => {
    const container = shallow(<Button children='test' ariaLabel='test' />);

    expect(container.find('button').length).toEqual(1);
  });

  it('should render one Child Div', () => {
    const container = shallow(
      <Button children='test' ariaLabel='test'>
        <div>Test</div>
      </Button>
    );

    expect(container.find('button').children().length).toEqual(1);
  });

  it('should render Loader Component if Loading', () => {
    const container = shallow(<Button children='test' loading ariaLabel='test' />);

    expect(container.contains(<Loading />)).toEqual(true);
  });

  it('should not render Loader Component if loading prop absent', () => {
    const container = shallow(<Button children='test' color='none' ariaLabel='test' />);

    expect(container.contains(<Loading />)).toEqual(false);
  });

  it('should render wrapped button if label passed', () => {
    const container = shallow(<Button children='test' label='test' ariaLabel='test' />);

    expect(container.find('.cui-button__container--small').length).toEqual(1);
  });

  it('should render wrapped button in large container if label and containerLarge passed', () => {
    const container = shallow(<Button children='test' label='test' containerLarge ariaLabel='test' />);

    expect(container.find('.cui-button__container').length).toEqual(1);
  });

  it('should be type button by default', () => {
    const container = shallow(<Button children='test' ariaLabel='test' />);

    expect(container.props().type).toEqual('button');
  });

  it('should show active class when passed active prop', () => {
    const container = shallow(<Button active children='test' ariaLabel='test' />);

    expect(container.find('.cui-button').hasClass('active')).toEqual(true);
  });

  it('should show type if passed one', () => {
    const container = shallow(<Button children='test' type='submit' ariaLabel='test' />);

    expect(container.props().type).toEqual('submit');
  });

  it('should output anchor if passed tag a', () => {
    const container = shallow(<Button children='test' tag='a' ariaLabel='test' />);

    expect(container.find('a').length).toEqual(1);
  });

  it('should handle disabled state', () => {
    const container = shallow(<Button children='test' disabled ariaLabel='test' />);

    expect(container.props().disabled).toEqual(true);
  });

  it('should handle onClick event', () => {
    let count = 0;
    const countUp = () => count++;
    const container = shallow(<Button children='test' onClick={countUp} ariaLabel='test' />);

    container.find('button').simulate('click');
    expect(count).toEqual(1);
  });

  it('should handle keyPress as onClick event', () => {
    let count = 0;
    const countUp = () => count++;
    const container = mount(<Button children='test' onClick={countUp} ariaLabel='test' />);

    container.find('button').simulate('click');
    container
      .find('button')
      .simulate('keyPress', { which: 13, charCode: 13, key: 'Space' });
    expect(count).toEqual(2);
  });
});
