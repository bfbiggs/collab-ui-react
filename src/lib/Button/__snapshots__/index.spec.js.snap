// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <Button /> should match SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    active={false}
    ariaLabel="test"
    ariaLabelledBy=""
    circle={false}
    className=""
    color=""
    containerLarge={false}
    disabled={false}
    expand={false}
    href=""
    label=""
    large={false}
    loading={false}
    onClick={null}
    style={Object {}}
    tag="button"
    type="button"
  >
    test
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "alt": "test",
      "aria-label": "test",
      "children": Array [
        false,
        <span
          className="cui-button__children"
          style={
            Object {
              "opacity": "1",
            }
          }
        >
          test
        </span>,
      ],
      "className": "cui-button",
      "disabled": false,
      "href": undefined,
      "onClick": null,
      "onKeyPress": [Function],
      "role": "",
      "style": Object {},
      "tabIndex": 0,
      "type": "button",
    },
    "ref": "button",
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": "child-1",
        "nodeType": "host",
        "props": Object {
          "children": "test",
          "className": "cui-button__children",
          "style": Object {
            "opacity": "1",
          },
        },
        "ref": null,
        "rendered": "test",
        "type": "span",
      },
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "alt": "test",
        "aria-label": "test",
        "children": Array [
          false,
          <span
            className="cui-button__children"
            style={
              Object {
                "opacity": "1",
              }
            }
          >
            test
          </span>,
        ],
        "className": "cui-button",
        "disabled": false,
        "href": undefined,
        "onClick": null,
        "onKeyPress": [Function],
        "role": "",
        "style": Object {},
        "tabIndex": 0,
        "type": "button",
      },
      "ref": "button",
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": "child-1",
          "nodeType": "host",
          "props": Object {
            "children": "test",
            "className": "cui-button__children",
            "style": Object {
              "opacity": "1",
            },
          },
          "ref": null,
          "rendered": "test",
          "type": "span",
        },
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
